README.md
# covid-19

Jupyter notebook built to digest, compile and analyze data related to the covid-19

The charts are displayed using plotly <https://plot.ly>

Run the jupyter 

* on Google Colab : [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ALT-F1/covid-19/blob/master/daily_geographic_distribution_of_covid_19_cases_worldwide.ipynb)

* on Azure Notebooks : <https://notebooks.azure.com/abdelkrim/projects/covid-19>


Author : Abdelkrim Boujraf working for <http://www.alt-f1.be>

# how to run the code using python3?

* download the code on your machine
* install python3
* run `pip install -r requirements.txt`
* run the code `python daily_geographic_distribution_of_covid_19_cases_worldwide.py`

## results in the browser

* your default browser will start and display three charts

### Covid-19 - New confirmed cases in Belgium and Italy

![Covid-19 - New confirmed cases in Belgium and Italy](./screenshots/2020-03-16/covid-19-belgium_and_italy_new_confirmed_cases.png "Covid-19 - New confirmed cases in Belgium and Italy")

### Covid-19 - New deaths in Belgium and Italy

![Covid-19 - New deaths in Belgium and Italy](./screenshots/2020-03-16/covid-19-belgium_and_italy_new_deaths.png "Covid-19 - New deaths in Belgium and Italy")

### Covid-19 - Worldwide, new confirmed cases and new deaths

![Covid-19 - Worldwide, new confirmed cases and new deaths](./screenshots/2020-03-16/covid-19-worldwide_new_confirmed_cases_and_deaths.png "Covid-19 - Worldwide, new confirmed cases and new deaths")


# how to run the code on a jupyter notebook

upload the jupyter notebook on Google Colab <https://colab.research.google.com/>

The jupyter notebook runs exactly the same code as the python code.

The user chooses the countries s.he wants to compare by using the 'Select multiple' widget

![Jupyter notebook - choose the countries you want to compare](./screenshots/2020-03-18/jupyter_notebook-the_user_can_choose_the_countries_to_compare.png "Jupyter notebook - choose the countries you want to compare")


# license

The GNU Affero General Public License is a modified version of the ordinary GNU GPL version 3. It has one added requirement: if you run a modified program on a server and let other users communicate with it there, your server must also allow them to download the source code corresponding to the modified version running there.

* why affero? https://www.gnu.org/licenses/why-affero-gpl.html

* source: https://www.gnu.org/licenses/agpl-3.0.en.html


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.0.1](https://github.com/ALT-F1/covid-19/compare/v2.0.0...v2.0.1) (2020-04-27)

## [2.0.0](https://github.com/ALT-F1/covid-19/compare/v1.2.7...v2.0.0) (2020-04-27)


### Features

* scatter containing several continents works! ([f88fa8d](https://github.com/ALT-F1/covid-19/commit/f88fa8ddab465a66b35b2cc6e9cf5c997e3b1649))

### [1.2.7](https://github.com/ALT-F1/covid-19/compare/v1.2.6...v1.2.7) (2020-04-27)


### Features

* create library building the DataFrame containing the data from the ECDC and display an animation on plotly ([99cb106](https://github.com/ALT-F1/covid-19/commit/99cb106fd70d7cb9d738eb58c9e156b7a5e83111))

### [1.2.6](https://github.com/ALT-F1/covid-19/compare/v1.2.5...v1.2.6) (2020-04-17)


### Features

* jupyter notebook drawing an animated scatter on COVID data ([32aa373](https://github.com/ALT-F1/covid-19/commit/32aa3737e3504b62cc4c8723d4b36ed12daf8014))

### [1.2.5](https://github.com/ALT-F1/covid-19/compare/v1.2.4...v1.2.5) (2020-04-17)


### Features

* scatter animation of the continents (but only one continent at a time) ([9a1992a](https://github.com/ALT-F1/covid-19/commit/9a1992adc15986631aa23f2b3371ca4df85c9322))

### [1.2.4](https://github.com/ALT-F1/covid-19/compare/v1.2.3...v1.2.4) (2020-04-08)

### [1.2.3](https://github.com/ALT-F1/covid-19/compare/v1.2.2...v1.2.3) (2020-03-29)


### Bug Fixes

* correct the name of the variables due to a change from the ECDC ([097d7f6](https://github.com/ALT-F1/covid-19/commit/097d7f6a7356362939cda9bd6675ca5619d954e1))

### [1.2.2](https://github.com/ALT-F1/covid-19/compare/v1.2.1...v1.2.2) (2020-03-21)


### Features

* check files from ECDC with 2 extensions: xls and xlsx ([f1d57b1](https://github.com/ALT-F1/covid-19/commit/f1d57b11403757a15aa48281d5dad76dd386b061))
* check files from ECDC with 2 extensions: xls and xlsx ([250deaa](https://github.com/ALT-F1/covid-19/commit/250deaa7f2ac67294b9f3a37308209f64a44cbec))

### [1.2.1](https://github.com/ALT-F1/covid-19/compare/v1.2.0...v1.2.1) (2020-03-18)

## [1.2.0](https://github.com/ALT-F1/covid-19/compare/v1.1.0...v1.2.0) (2020-03-18)

## [1.1.0](https://github.com/ALT-F1/covid-19/compare/v1.0.0...v1.1.0) (2020-03-17)


### Features

* adapt the breaking changes from the v1.0.0 to the jupyter notebook ([1b82693](https://github.com/ALT-F1/covid-19/commit/1b826935174644c1463b8e201e32d30e240866eb))

## [1.0.0](https://github.com/ALT-F1/covid-19/compare/v0.0.1...v1.0.0) (2020-03-17)


### ⚠ BREAKING CHANGES

* ECDC has changed the format of the file and its filename

data source from the ECDC is renamed *.xls into *.xlsx

renamed fields: CountryExp -> Countries and territories, NewConfCases -> Cases, Newdeaths -> Deaths

removed fields: Gaul1Nuts1, EU

### Features

* search the for latest file available from the ECDC ([63056a3](https://github.com/ALT-F1/covid-19/commit/63056a3b367accffabf16a778cdea60a6cfe929f))

### 0.0.1 (2020-03-17)


### Features

* Display charts about the geographic distribution of COVID-19 cases worldwide ([3cba906](https://github.com/ALT-F1/covid-19/commit/3cba9069cdb8aa301939cbd3397bdb886eccbc8c))


plotly/covid.py
from country_continent_codes import CountryContinentCodes
from datetime import datetime, timedelta
import inspect
import logging
import numpy as np
import pandas as pd
import urllib.error
import urllib.request

"""# Load the LATEST data from the European Center for Disease Prevention and Control (ECDC)"""


"""
    helpers
"""


class Utils:

    """
        check if an url exists
    """
    @classmethod
    def check_if_data_source_exists(cls, url):
        url_exists = False
        try:
            conn = urllib.request.urlopen(url)
        except urllib.error.HTTPError as e:
            # Return code error (e.g. 404, 501, ...)
            # ...
            # print('HTTPError: {}, the file does not exist'.format(e.code))
            url_exists = False
        except urllib.error.URLError as e:
            # Not an HTTP-specific error (e.g. connection refused)
            # ...
            print('URLError: {}'.format(e.reason))
            url_exists = False
        else:
            # 200
            # ...
            print(f'Data source is : {url}')
            url_exists = True
        return (url_exists), url


class ECDC:

    def __init__(self):
        logging.info(f"{'='*25} {inspect.currentframe().f_code.co_name}")

        # load the data
        self.df = self.extract_datasource()
        self.df = self.transform_add_columns(self.df)
        
    def info(self):
        logging.info(f"{'='*25} {inspect.currentframe().f_code.co_name}")
        print(f"{self.df['continent'].unique()}")
    """
        extract transform and load the data from the country continent codes list
    """
    def load_country_continent_codes(self):
        logging.info(f"{'='*25} {inspect.currentframe().f_code.co_name}")

        return CountryContinentCodes().cleanse(), CountryContinentCodes().map()

    """
        Add a column 'continent' mapping the country with its continent
        Add columns 'cases_sum_to_date', 'deaths_sum_to_date', 'cases_growth'
    """

    def transform_add_columns(self, df):
        logging.info(f"{'='*25} {inspect.currentframe().f_code.co_name}")
        self.country_continent_codes_df, self.alpha3_region_map = self.load_country_continent_codes()
        df['continent'] = df['countryterritoryCode'].map(
            self.alpha3_region_map)
        df = self.add_columns_with_growth(df, periods=1)
        df['dateRep_str'] = df["dateRep"].astype(str)
        return df

    def add_columns_with_growth(self, df, periods=1):
        logging.info(f"{'='*25} {inspect.currentframe().f_code.co_name}")
        countries =  df['countriesAndTerritories'].unique()
        print(f"uniques countries : {len(countries)}")
        
        for country in countries:
            df_growth = df[
                df["countriesAndTerritories"] == country].copy(deep=True)
            df_growth = df_growth.sort_values(
                by=["dateRep"],
                ascending=True
            )
            df_growth['cases_sum_to_date'] = df_growth['cases'].cumsum()
            df_growth['deaths_sum_to_date'] = df_growth['deaths'].cumsum()
            df_growth['cases_growth'] = df_growth['cases_sum_to_date'].pct_change(
                periods=periods, fill_method='pad')
            df_growth['deaths_growth'] = df_growth['deaths_sum_to_date'].pct_change(
                periods=periods, fill_method='pad')
            df = df.append(df_growth,ignore_index=True)
        
        # remove all lines containing infinities to 100 and "Not A Number" (NaN) 
        # df = df.replace([np.inf, -np.inf], np.nan)
        df = df.replace([np.inf, -np.inf], 100)
        df = df.dropna()
        
        return df
    """
        load the data from the ECDC
    """

    def extract_datasource(self):
        logging.info(f"{'='*25} {inspect.currentframe().f_code.co_name}")

        datasource_url_template = "https://www.ecdc.europa.eu/sites/default/files/documents/COVID-19-geographic-disbtribution-worldwide-%latest_datasource_date%"

        datasource_url = self.find_latest_datasource(datasource_url_template)

        if datasource_url is None:
            raise Exception(
                f'There is no datasource on the ECDC website. is https://www.ecdc.europa.eu down?')

        # read the data from the excel sheet
        return pd.read_excel(datasource_url)

    """
        find the latest datasource starting today from the ECDC website

        format of the url: 
            https://www.ecdc.europa.eu/sites/default/files/documents/COVID-19-geographic-disbtribution-worldwide-2020-03-17.xlsx

            https://www.ecdc.europa.eu/sites/default/files/documents/COVID-19-geographic-disbtribution-worldwide-%latest_datasource_date%.xlsx

    """

    def find_latest_datasource(self, url):
        is_found = False
        i = -1
        while is_found == False and i < 5:
            i += 1
            latest_datasource_date = (
                datetime.now() - timedelta(i)).strftime('%Y-%m-%d')
            datasource_url = url.replace(
                '%latest_datasource_date%',
                latest_datasource_date
            )
            # print(f'Search for this latest date: {latest_datasource_date}',
            #      f'datasource_url: {datasource_url}')
            is_found, existing_datasource_date = Utils.check_if_data_source_exists(
                datasource_url + ".xls"
            )
            if is_found == False:
                is_found, existing_datasource_date = Utils.check_if_data_source_exists(
                    datasource_url + ".xlsx"
                )

        if is_found == True:
            print(f'Latest date is : {latest_datasource_date}')
            return existing_datasource_date
        else:
            return None

    def data(self):
        return self.df


class DATA:
    def __init__(self):
        logging.info(f"{'='*25} {inspect.currentframe().f_code.co_name}")
        self._ecdc = ECDC()

    def ecdc(self):
        logging.info(f"{'='*25} {inspect.currentframe().f_code.co_name}")
        return self._ecdc.df


class COVID:
    def __init__(self):
        self.data = DATA()


if __name__ == "__main__":
    covid = COVID()
    print(f"ECDC DataFrame head:\n{covid.data.ecdc().head(3)}")
    print(f"ECDC DataFrame tail:\n{covid.data.ecdc().tail(3)}")


plotly/animation.py
import dash
import dash_core_components as dcc
import dash_html_components as html
import numpy as np
import plotly.graph_objects as go  # or plotly.express as px
import plotly.express as px
import locale
from covid import COVID
locale.setlocale(locale.LC_ALL, '')

def build_df():
    covid = COVID()
    return covid.data.ecdc()

def plot_scatter(df, continent="Americas",start_date="2020-03-01"):

    mask = (df['dateRep'] > start_date)
    df = df.loc[mask]


    print(f"before sort: {df.shape}")
    df = df.sort_values(by=["dateRep"], ascending=True)
    #df = df[(df['continent'] == continent)]
    print(f"unique continents: {df['continent'].unique()}")
    return px.scatter(
        df,
        x="cases_sum_to_date", y="cases_growth",
        animation_frame="dateRep_str", animation_group="countriesAndTerritories",
        size="deaths_sum_to_date", color="continent", hover_name="countriesAndTerritories",
        log_x=True, size_max=50,
        range_x=[df['cases_sum_to_date'].min(), df['cases_sum_to_date'].max()],
        range_y=[df['cases_growth'].min(), df['cases_growth'].quantile(0.75)]
    )

df = build_df()

fig = plot_scatter(df)
#fig = go.Figure()  # or any Plotly Express function e.g. px.bar(...)
# fig.add_trace( ... )
# fig.update_layout( ... )


app = dash.Dash()
app.layout = html.Div([
    dcc.Graph(figure=fig)
])

# Turn off reloader if inside Jupyter
app.run_server(debug=True, use_reloader=False)


plotly/country_continent_codes.py
# -*- coding: utf-8 -*-

# Copyright (c) ALT-F1 SPRL, Abdelkrim Boujraf. All rights reserved.
# Licensed under the EUPL License, Version 1.2. See LICENSE in the project root for license information.

import pandas as pd
import numpy as np

class CountryContinentCodes:

    def __init__(self):
        self.df = pd.read_csv(
            'https://raw.githubusercontent.com/lukes/ISO-3166-Countries-with-Regional-Codes/master/all/all.csv',
            sep=',',
            usecols=['alpha-3', 'region'],
            #dtype={"region": str, "alpha-3": str},
            # dtype=str
            index_col='alpha-3'
        )

    def data(self):
        return self.df

    """
        clease the data : 
        - drop the NaN from the DataFrame
    """
    def cleanse(self):
        df = self.df.dropna()
        return df

    """
        create a dictionnary including alpha3:region (continent)
        Useful to create a column with the region/continents based on a alpha-3 code
    """
    def map(self):
        return self.df['region'].to_dict()

daily_geographic_distribution_of_covid_19_cases_worldwide.py
# -*- coding: utf-8 -*-
"""daily_geographic_distribution_of_covid-19_cases_worldwide-v0.0.7.ipynb

Automatically generated by Colaboratory.

Original file is located at
    https://colab.research.google.com/drive/1X6lSPpBi-SKuR0kA9mLET5aFwDi4am8c

# Display charts about the geographic distribution of COVID-19 cases worldwide

> Data source is "today’s data on the geographic distribution of COVID-19 cases worldwide": http://bit.ly/2U9YcGS

> ECDC Dashboard : http://bit.ly/38Ysgum

# Press CTRL+F9 (Runtime > Run all in the menu) to execute the notebook

> Edit the line below with the countries you would like to analyze
"""

countries = ['China', 'Morocco', 'Belgium', 'Spain', 'France', 'Italy', 'United_States_of_America']

"""> Execute the notebook (CTRL-F9)

> Scroll down until the charts
"""

# setup the python program
import pandas as pd
import numpy as np
from plotly.offline import iplot
import plotly.graph_objs as go
from datetime import datetime, timedelta
import locale
import urllib.request, urllib.error
import ipywidgets as widgets
locale.setlocale(locale.LC_ALL, '')

"""# Load the LATEST data from the European Center for Disease Prevention and Control (ECDC)"""

def check_if_data_source_exists(url):
  """
  check if an url exists
  """
  url_exists = False
  try:
    conn = urllib.request.urlopen(url)
  except urllib.error.HTTPError as e:
      # Return code error (e.g. 404, 501, ...)
      # ...
      print('HTTPError: {}, the file does not exist'.format(e.code))
      url_exists = False
  except urllib.error.URLError as e:
      # Not an HTTP-specific error (e.g. connection refused)
      # ...
      print('URLError: {}'.format(e.reason))
      url_exists = False
  else:
      # 200
      # ...
      print('The url exists')
      url_exists = True
  return (url_exists), url

def find_latest_datasource(url):
  """
  find the latest datasource starting today from the ECDC website

  format of the url: 
    https://www.ecdc.europa.eu/sites/default/files/documents/COVID-19-geographic-disbtribution-worldwide-2020-03-17.xlsx
    https://www.ecdc.europa.eu/sites/default/files/documents/COVID-19-geographic-disbtribution-worldwide-%latest_datasource_date%.xlsx

  """
  is_found=False
  i=-1
  while is_found == False and i<5:
    i += 1
    latest_datasource_date = (datetime.now() - timedelta(i)).strftime('%Y-%m-%d')
    datasource_url = url.replace('%latest_datasource_date%', latest_datasource_date)
    print(f'Search for this latest date: {latest_datasource_date}',
          f'datasource_url: {datasource_url}')
    is_found, existing_datasource_date = check_if_data_source_exists(datasource_url + ".xls")
    if is_found == False:
      is_found, existing_datasource_date = check_if_data_source_exists(datasource_url + ".xlsx")

  if is_found == True:
    print(f'Download the data until : {latest_datasource_date}')
    return existing_datasource_date
  else:
    return None

# load the data
datasource_url_template = "https://www.ecdc.europa.eu/sites/default/files/documents/COVID-19-geographic-disbtribution-worldwide-%latest_datasource_date%"
if datasource_url_template is None:
  print(f'There is no datasource on the ECDC website. is https://www.ecdc.europa.eu down?')
  exit
datasource_url = find_latest_datasource(datasource_url_template)

df = pd.read_excel(datasource_url)

df.columns

"""# What is the range of data that has been collected from the ECDC?"""

start_date = df.dateRep.min().strftime("%b %d %Y")
end_date = df.dateRep.max().strftime("%b %d %Y")

print(
    f'DATE RANGE:\n',
    f'start date : {start_date}\n', 
    f'end date : {end_date}'
      )

"""# Display Totals : Confirmed   and Deaths"""

cv19_countries_day = df.groupby(by=['dateRep','countriesAndTerritories']).sum()[['cases','deaths']]
Total_confirmed = cv19_countries_day.groupby('dateRep').sum()[['cases','deaths']].sum()['cases']
Total_deaths = cv19_countries_day.groupby('dateRep').sum()[['cases','deaths']].sum()['deaths']
print(f"{end_date}:\n Total Confirmed cases: {Total_confirmed:n}\n Total deaths: {Total_deaths:n}")

"""# Group the data by daily-New Confirmed cases (WW)"""

cv19_countries_day_new_confirmed_cases = df.groupby(by=['dateRep'], as_index=False)['cases'].sum()
print(f'WW New Confirmed cases first-5 records:\n\n {cv19_countries_day_new_confirmed_cases.head(5)}')
print(f'\nWW New Confirmed cases last-5 records:\n\n {cv19_countries_day_new_confirmed_cases.tail(5)}')

"""# Group the data by daily Deaths (WW)"""

cv19_countries_day_new_deaths = df.groupby(by=['dateRep'], as_index=False)['deaths'].sum()
print(f'WW New deaths first-5 records:\n\n {cv19_countries_day_new_deaths.head(5)}')
print(f'nWW New deaths last-5 records:\n\n {cv19_countries_day_new_deaths.tail(5)}')

"""# Draw a chart with daily New (WW) Confirmed Cases and"""

data = [
        go.Scatter(
        name="WW New Confirmed cases",
        x=cv19_countries_day_new_confirmed_cases['dateRep'],
        y=cv19_countries_day_new_confirmed_cases['cases']
        ),
        go.Scatter(
        name="WW New deaths",
        x=cv19_countries_day_new_deaths['dateRep'],
        y=cv19_countries_day_new_deaths['deaths']
        ),
]

# plot titles and axis labels
layout = go.Layout(
    title=f'Worldwide New Confirmed cases and deaths',
    xaxis=dict(
        title='Date'
    ),
    yaxis=dict(
        title='New Confirmed cases/New deaths'
    )
)

# plot the Figure object using the 'iplot' method
fig = go.Figure(data=data, layout=layout)
iplot(fig)

"""# ACTION : Choose the countries you want to compare or update the list of countries you want to study (Multi-selection with CTRL key)"""

countries_widget = widgets.SelectMultiple(
    options=df['countriesAndTerritories'].unique().tolist(),
    value=countries,
    rows=10,
    description='countriesAndTerritories',
    disabled=False
)

try:
  display(countries_widget)
  #print('you are running a jupyter notebook')
except:
  print('you are not running a jupyter notebook')

df.head(2)

if countries_widget is not None:
    if countries_widget.value is not None:
      countries = countries_widget.value
      
print(f'Selected countries : {countries}')

def display_bar_chart(countries, 
                      bar_name="New confirmed cases in", 
                      bar_x='dateRep', 
                      bar_y='cases',
                      layout_x_title='Date', 
                      layout_y_title='New confirmed cases'
                      ):
  # use the DataFrame columns for generating data

  data = []
  for country in countries:
    # print(f"{country}")
    data.append(
        go.Bar(
          name=f"{bar_name} {country}",
          x=df[df['countriesAndTerritories']==country][bar_x],
          y=df[df['countriesAndTerritories']==country][bar_y]
        )
      )
  # plot titles and axis labels
  layout = go.Layout(
      title=f'{bar_name} {countries}',
      xaxis=dict(
          title=layout_x_title
      ),
      yaxis=dict(
          title=layout_y_title
      )
  )

  # plot the Figure object using the 'iplot' method
  fig = go.Figure(data=data, layout=layout)
  iplot(fig)

"""# Draw a chart including New Confirmed Cases in Italy and Belgium

## Some hints regarding the dynamic charts

> functions are available on the top-left of the screen. Click on the buttons

> the user can zoom in/zoom out

> click on a legend e.g., new confirmed cases (on the right) and focus on one single dataset

> click-and-drag your mouse to zoom on a dedicated area

> double click on the chart to reset the zoom
"""

df[df['countriesAndTerritories']=="China"].head(2)

display_bar_chart(
    countries, 
    bar_name="New confirmed cases in", 
    bar_x='dateRep', 
    bar_y='cases',
    layout_x_title='Date', 
    layout_y_title='New confirmed cases'
    )
display_bar_chart(
    countries,
    bar_name="New deaths in", 
    bar_x='dateRep', 
    bar_y='deaths',
    layout_x_title='Date', 
    layout_y_title='New deaths'
    )

"""# Are they Correlations between the fields 'deaths' and 'New cases'?"""

def display_correlation(df, selected_citizenship= 'Belgian', method='pearson'):
  correlation = df['cases'].corr(df['deaths'], method=method)
  print(f"correlation using '{method}'' method between '{selected_citizenship}' 'New cases' and 'deaths': {correlation}")

for country in countries:
  display_correlation(df[df['countriesAndTerritories']==country], country, method='pearson')

"""# CHANGELOG

v.0.0.7

> rename names due to changes in the raw data by the ECDC

v.0.0.2

> fix: ECDC has changed the format of the file

  > filename is renamed *.xlsx into *.xls

> feat: add widget to select the countries you want a analysis about

> refactor: one function displays diverse bart charts

v.0.0.1 
  
> fix: ECDC has changed the format of the file

    > filename is renamed *.xls into *.xlsx
  
    > renamed fields: CountryExp -> Countries and territories, NewConfCases -> Cases, Newdeaths -> Deaths
  
    > removed fields: Gaul1Nuts1, EU

v.0.0.0 

    > feat: display charts for the World, Belgium and Italy

# LICENSE

(c) 2010-2020 Abdelkrim Boujraf, alt-f1 sprl <http://www.alt-f1.be>, Abdelkrim Boujraf <http://www.alt-f1.be/literature.html>

This work is licensed under a Creative Commons Attribution 4.0 International License: <http://creativecommons.org/licenses/by/4.0>

* Share — copy and redistribute the material in any medium or format 
* Adapt — remix, transform, and build upon the material for any purpose, even commercially

Vous êtes autorisé à <https://creativecommons.org/licenses/by/4.0/deed.fr>: 

* Partager : copier, distribuer et communiquer le matériel par tous moyens et sous tous formats
* Adapter : remixer, transformer et créer à partir du matériel pour toute utilisation, y compris commerciale
"""